File: Cracking the Coding Interview/Array_Strings/PermutationInString.java
public class PermutationInString {
    public boolean checkInclusion(String s1, String s2) {
        if(s1.length() > s2.length()) return false;

        int[] s1Counter = new int[26];
        int[] s2Counter = new int[26];

        for(int i = 0; i < s1.length(); i++) {
            s1Counter[s1.charAt(i)-'a'] += 1;
            s2Counter[s2.charAt(i)-'a'] += 1;
        }

        int matches = 0;

        for(int i=0; i<26; i++) {
            if(s1Counter[i] == s2Counter[i]) matches++;
        }

        int left = 0;

        for(int i = s1.length(); i < s2.length(); i++) {
            if(matches == 26) return true;
            
            int index = s2.charAt(i) - 'a';
            s2Counter[index] += 1;

            if(s1Counter[index] == s2Counter[index]) {
                matches++;
            } else if(s1Counter[index]+1 == s2Counter[index]) {
                matches--;
            }

            index = s2.charAt(left) - 'a';

            s2Counter[index] -= 1;
            if(s1Counter[index] == s2Counter[index]) {
                matches++;
            } else if(s1Counter[index]-1 == s2Counter[index]) {
                matches--;
            }
            left++;
        }

        return matches == 26;
    }
}

File: NeetCode/SlidingWindow/PermutationInString.py
"""
LeetCode: 567. Permutation in String
NeetCode: 18/150

Given two strings s1 and s2, return true if s2 contains a permutation of s1, or false otherwise.

In other words, return true if one of s1's permutations is the substring of s2.
"""

class Solution:
    def checkInclusion(self, s1: str, s2: str) -> bool:
        if(len(s1) > len(s2)): return False
    
        s1_count, s2_count = [0]*26, [0]*26
        matches = 0
        
        for i in range(len(s1)):
            s1_count[ord(s1[i]) - ord('a')] += 1
            s2_count[ord(s2[i]) - ord('a')] += 1
            
        for i in range(26):
            matches += (1 if s1_count[i] == s2_count[i] else 0)
        
        left = 0
        
        for right in range(len(s1), len(s2)):
            if(matches == 26):
                return True
            
            index = ord(s2[right]) - ord('a')
            s2_count[index] += 1
            if(s1_count[index] == s2_count[index]):
                matches += 1
            elif(s1_count[index]+1 == s2_count[index]):
                matches -= 1
                
            index = ord(s2[left]) - ord('a')
            s2_count[index] -= 1
            if(s1_count[index] == s2_count[index]):
                matches += 1
            elif(s1_count[index]-1 == s2_count[index]):
                matches -= 1
            left += 1
            
        
        return matches == 26
